### ISSUE TITLE
[Maintainer quality of life improvement] - Add rudimentary support for Nix. 

### ISSUE THREAD

**Issue by isubasinghe** [argoproj/argo-workflows#10998]
# Summary

The dependencies in Argo Workflows are complex and it is hard to get a reproducible build and in my past experience it was hard to get a build at all. 
The Makefile is heavily opinionated and takes on the role of a task runner, build tool and a package manager. Migrating some of it's functionality away to a more reproducible setup will make developing easier for developers. 

# Use Cases
I would use this as my daily development environment. 

I actually have an existing flake now in our [fork](https://github.com/pipekit/argo-workflows/pull/21), I spent the day porting it over to the upstream repo, so not only will this work help us, it should help even more developers to get started on Argo. 

# Why Nix? 
https://serokell.io/blog/what-is-nix
TL;DR 
The claims (which I also agree with are) are
  1. That the Filesystem Hierarchy Standard is fundamentally incompatible with reproducibility 
  2. Docker is a band-aid over this.
